### DISCUSSION TITLE
Conversion from/to numpy array creates weird edges

### DISCUSSION THREAD

**Discussion by Ripper346** [networkx/networkx#4692] [Q&A] [answered]
Hi, I don't know if I am just too stupid or I have encountered a bug. It is my first project in which I use NetworkX. At some point I load a dataset from TUDataset using their read method to have a dataset of networkX graphs. Other than nodes and edges, it set some attributes on the graph, nodes and edges. Then I need to reorder the nodes of every graph and reconstruct a new networkX graph. For this purpose I wrote two methods that, for now, I call one after the other:

```python
def graph_to_numpy_matrix(
    graph: nx.Graph, sort_nodes_list: List[int] = None
) -> Tuple[np.ndarray, Dict, List[Dict], Dict[Tuple[int, int], Dict]]:
    mat = nx.to_numpy_array(graph, nodelist=sort_nodes_list)
    graph_attr = graph.graph
    node_attr = np.array([graph.nodes[n] for n in graph.nodes])
    edge_attr = np.array([graph.edges[e] for e in graph.edges])
    if sort_nodes_list is not None:
        node_attr = node_attr[sort_nodes_list]
        edge_attr = {(sort_nodes_list[e[0]], sort_nodes_list[e[1]]): edge_attr[idx]
                     for idx, e in enumerate(graph.edges)}
    return mat, graph_attr, node_attr, edge_attr

def numpy_matrix_to_graph(
    mat: np.ndarray,
    graph_attr: Dict = None,
    node_attr: List[Dict] = None,
    edge_attr: Dict[Tuple[int, int], Dict] = None
) -> nx.Graph:
    graph = nx.from_numpy_array(mat)
    if graph_attr:
        graph.graph = graph_attr
    if node_attr is not None:
        for idx in range(len(graph.nodes)):
            for key in node_attr[idx].keys():
                graph.nodes[idx][key] = node_attr[idx][key]
    if edge_attr is not None:
        for e in graph.edges:
            for key in edge_attr[e].keys():
                graph.edges[e][key] = edge_attr[e][key]
    return graph
```

However, the recreation of the graph doesn't work. When it recreates the edge attributes, it gives a KeyError that it doesn't exist a tuple as key in `edge_attr` so it doesn't have a specific edge.

The culprit should be `edge_attr = {(sort_nodes_list[e[0]], sort_nodes_list[e[1]]): edge_attr[idx] for idx, e in enumerate(graph.edges)}` but I can't found the error: `sort_nodes_list` has always unique elements from 0 to len(nodes) - 1 as expected. Using it I can map the new places of the nodes and as a consequence the new edges coordinates. In the middle I don't do anything else. 

Thinking about something strange in `sort_nodes_list` I tried only to invert the order of the nodes with a `sort_nodes_list = list(range(len(graph.nodes) - 1, -1, -1))` but it still gave the error (for another key/edge).

Does anyone see the mistake, something I forgot or is there something wrong with the conversion to/from numpy array of the graph? Thank you.

**Comment by rossbar** [comment]
One thing to keep in mind about round-tripping from a graph<->matrix representation is that the nodes themselves are not necessarily preserved, only the relationships between them. For example:

```python
>>> G = nx.Graph([(0, 2)])
>>> G.nodes()
NodeView((0, 2))
>>> H = nx.from_numpy_array(nx.to_numpy_array(G))
>>> H.nodes()  # The node 2 in G has been remapped to 1 in H
NodeView((0, 1))
```

I'm not 100% this is the source of your problem, but it seems like a good place to start investigating.

**Comment by MridulS** [comment, marked as answer]
Hi, the issue here is the *direction* of edges in the `edge_attr` dictionary.
``` python
for e in graph.edges:
  for key in edge_attr[e].keys():
      graph.edges[e][key] = edge_attr[e][key]
```
`graph.edges` may give out an edge like `(u, v)` and the function `graph_to_numpy_matrix` may return a `edge_attr` dictionary with entry only for `(v, u)`.
``` python
In [7]: d = {(0, 1):  "I am here"}

In [8]: d[(0, 1)]
Out[8]: 'I am here'

In [9]: d[(1, 0)]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-59-7de873ef528d> in <module>
----> 1 d[(1, 0)]

KeyError: (1, 0)
```

> reorder the nodes of every graph and reconstruct a new networkX graph

I am curious about this, what's the reason behind the reorder. Order shouldn't make any difference in terms of a graph here.
The reconstructed graph would be isomorphic to the original graph.

**Reply by Ripper346** [reply]
Oh, thank you both, especially @MridulS, I don't know why I thought they were redundant.

About your curiosity, I know that it doesn't make any difference as it is, it is part of bigger program, I am trying to do an augmentation algorithm by dividing the graphs in clusters and swap them between graphs, so I need to reorder the nodes according to the cluster classification. After I use the methods I really need only the adjacency matrix but doing also the reorder of the graph removes a lot of troubles then to align data and results. I have almost finished it, I am only missing the saving of the extra labels of the graphs to use the dataset to a deep learning algorithm
